File: cpp/src/memory.cpp
#include "memory.h"
#include "cartridge.h"
#include "ppu.h"
#include "apu.h"

Memory::Memory() {
    ram.fill(0);
}

uint8_t Memory::read(uint16_t addr) {
    if (addr < 0x2000) {
        return ram[addr & 0x7FF];
    } else if (addr < 0x4000) {
        return readPPU(addr);
    } else if (addr < 0x4020) {
        return readAPU(addr);
    } else if (addr < 0x6000) {
        return 0;
    } else if (addr < 0x8000) {
        if (cartridge) {
            return cartridge->readPRG(addr);
        }
        return 0;
    } else {
        if (cartridge) {
            return cartridge->readPRG(addr);
        }
        return 0;
    }
}

void Memory::write(uint16_t addr, uint8_t value) {
    if (addr < 0x2000) {
        ram[addr & 0x7FF] = value;
    } else if (addr < 0x4000) {
        writePPU(addr, value);
    } else if (addr < 0x4020) {
        writeAPU(addr, value);
    } else if (addr < 0x6000) {
        // Open bus
    } else if (addr < 0x8000) {
        if (cartridge) {
            cartridge->writePRG(addr, value);
        }
    } else {
        if (cartridge) {
            cartridge->writePRG(addr, value);
        }
    }
}

uint16_t Memory::readWord(uint16_t addr) {
    uint8_t lo = read(addr);
    uint8_t hi = read(addr + 1);
    return (hi << 8) | lo;
}

void Memory::writeWord(uint16_t addr, uint16_t value) {
    write(addr, value & 0xFF);
    write(addr + 1, (value >> 8) & 0xFF);
}

void Memory::setCartridge(std::shared_ptr<Cartridge> cartridge) {
    this->cartridge = cartridge;
}

void Memory::setPPU(std::shared_ptr<class PPU> ppu) {
    this->ppu = ppu;
}

void Memory::setAPU(std::shared_ptr<class APU> apu) {
    this->apu = apu;
}

uint8_t Memory::readPPU(uint16_t addr) {
    if (ppu) {
        return ppu->read(addr);
    }
    return 0;
}

void Memory::writePPU(uint16_t addr, uint8_t value) {
    if (ppu) {
        ppu->write(addr, value);
    }
}

uint8_t Memory::readAPU(uint16_t addr) {
    if (apu) {
        return apu->read(addr);
    }
    return 0;
}

void Memory::writeAPU(uint16_t addr, uint8_t value) {
    if (apu) {
        apu->write(addr, value);
    }
}

File: android/emulator/src/main/java/emulator/Memory.kt
package com.nes.emulator

/**
 * Gerencia toda a memória do NES, incluindo RAM, ROM e registradores.
 * Arquitetura de memória:
 * 0x0000-0x07FF: RAM interna (2KB)
 * 0x0800-0x0FFF: Espelho de RAM
 * 0x1000-0x1FFF: Espelho de RAM
 * 0x2000-0x2007: Registradores PPU
 * 0x2008-0x3FFF: Espelho de registradores PPU
 * 0x4000-0x4017: Registradores APU e entrada
 * 0x4018-0x401F: Teste de ROM
 * 0x4020-0x5FFF: Cartridge RAM
 * 0x6000-0x7FFF: Cartridge RAM
 * 0x8000-0xFFFF: Cartridge ROM
 */
class Memory(val cartridge: Cartridge) {
    private val ram = ByteArray(0x800)  // 2KB RAM interna
    private val ppuRegisters = ByteArray(8)
    private val apuRegisters = ByteArray(0x18)
    
    var ppuAddress: Int = 0
    var ppuData: Int = 0
    var ppuScroll: Int = 0
    var ppuControl: Int = 0
    var ppuStatus: Int = 0
    
    var apuStatus: Int = 0
    var apuFrameCounter: Int = 0
    
    var controllerData: Int = 0
    var controllerStrobe: Boolean = false
    
    fun read(address: Int): Int {
        return when {
            address < 0x2000 -> ram[address and 0x7FF].toInt() and 0xFF
            address < 0x4000 -> {
                when (address and 0x7) {
                    2 -> ppuStatus
                    4 -> 0  // OAM data
                    7 -> ppuData
                    else -> 0
                }
            }
            address < 0x4020 -> {
                when (address) {
                    0x4015 -> apuStatus
                    0x4016 -> controllerData
                    0x4017 -> 0
                    else -> 0
                }
            }
            address < 0x6000 -> cartridge.readRam(address)
            address < 0x8000 -> cartridge.readRam(address)
            else -> cartridge.readRom(address)
        }
    }
    
    fun write(address: Int, value: Int) {
        when {
            address < 0x2000 -> ram[address and 0x7FF] = value.toByte()
            address < 0x4000 -> {
                when (address and 0x7) {
                    0 -> ppuControl = value
                    1 -> {}  // PPU mask
                    3 -> {}  // OAM address
                    4 -> {}  // OAM data
                    5 -> ppuScroll = value
                    6 -> ppuAddress = (ppuAddress and 0xFF) or (value shl 8)
                    7 -> ppuData = value
                }
            }
            address < 0x4020 -> {
                when (address) {
                    0x4014 -> {}  // OAM DMA
                    0x4015 -> apuStatus = value
                    0x4016 -> controllerStrobe = (value and 1) != 0
                    0x4017 -> apuFrameCounter = value
                    else -> if (address < 0x4018) apuRegisters[address - 0x4000] = value.toByte()
                }
            }
            address < 0x6000 -> cartridge.writeRam(address, value)
            address < 0x8000 -> cartridge.writeRam(address, value)
            else -> cartridge.writeRom(address, value)
        }
    }
    
    fun readWord(address: Int): Int {
        val low = read(address)
        val high = read(address + 1)
        return (high shl 8) or low
    }
    
    fun writeWord(address: Int, value: Int) {
        write(address, value and 0xFF)
        write(address + 1, (value shr 8) and 0xFF)
    }
}
